### PR TITLE
Add statsd metrics reporting

### PR DESCRIPTION
**Pull request by ChimeraCoder**
Add statsd metrics reporting. Specifically, we want to record the size of each line processed, along with the box type and the logfile name. (We specifically want box *type* and not hostname to avoid problems with tag cardinality).

The easiest way to do this is to send to the local veneur on each box. I've included the default as a constant, but we could move that to puppet if we really wanted (though it'll always be the same).

Note that this will require bumping to Go 1.7+, because of the use of subtests. We could scale that back to Go 1.5+, with `GO15VENDOREXPERIMENT=1`, but I don't think we'd gain much - if anything - from doing that. We're already running 1.7 on veneur (and some other services) and 1.6+ on everything else.

r? @nelhage 

cc @kiran @tummychow @gphat

### CODE REVIEW

File: hostname.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	"fmt"
+	"regexp"
+)
+
+var hostnameRegex = regexp.MustCompile(`^([A-Za-z]+)`)
+
+// ParseHostname parses a hostname and returns the
+// host type (e.g. mybox1 and mybox-123.local will both
+// return 'mybox').
+func ParseHostname(hostname string) (string, error) {
```
Comment (nelhage) [inline hostname.go]: unilog is open-source and not necessarily Stripe specific; would making the tag an argument and plumbing it through internally make more sense? It'd also avoid parsing hostnames in random application code, which is a desirable property to reduce coupling to our infrastructure.

File: unilog.go [hunk 2]
```diff
@@ -126,7 +133,11 @@ func readlines(in io.Reader, bufsize int, shutdown chan struct{}) (<-chan string
 		for err == nil {
 			s, err = r.ReadString('\n')
 			if s != "" {
-				linec <- strings.TrimRight(s, "\n")
+				s = strings.TrimRight(s, "\n")
+				linec <- s
+				if stats != nil {
+					stats.Gauge("unilog.bytes", float64(len(s)), nil, 1.0)
```
Comment (nelhage) [inline unilog.go]: Do you really want a rate of 1.0 here? We log a lot of lines.

Comment (gphat) [inline unilog.go]: Should be a `Count` rather than `Gauge`?

Comment (ChimeraCoder) [inline unilog.go]: Yes, this should be a Count, not a gauge.

On that note (and the previous one) - we discussed in Slack what our approach to tagging should be, given that (a) tag cardinality is an issue, (b) we plan on being able to forward counters but can't yet, and (c) hostname may not always be the most meaningful way to identify what we care about (e.g. in a hypothetical containerized world).

So, our plan is to use `--statstag` and (for now) pass in the box type. (In the future, we could get creative with using the service and/or team in the metric name to reduce cardinality and then use the full hostname in the tag, but we don't have a unified naming scheme for metrics (yet) and given a problematic box type, it's easy enough to query splunk to see which host(s) are causing a problem.

**Comment by ChimeraCoder** [comment]
Copying for visiblity after rebase:


> We discussed in Slack what our approach to tagging should be, given that (a) tag cardinality is an issue, (b) we plan on being able to forward counters but can't yet, and (c) hostname may not always be the most meaningful way to identify what we care about (e.g. in a hypothetical containerized world).
> 
> So, our plan is to use --statstag and (for now) pass in the box type. (In the future, we could get creative with using the service and/or team in the metric name to reduce cardinality and then use the full hostname in the tag, but we don't have a unified naming scheme for metrics (yet) and given a problematic box type, it's easy enough to query splunk to see which host(s) are causing a problem.

**Comment by ChimeraCoder** [comment]
re-r? @nelhage 

(will squash the last commits before merging)

**Comment by nelhage** [comment]
lgtm

**Pull request merged**